Added producer json message validation

bazen-teklehaymanot opened the pull request:
#2

elchinmemphis commented on src/Memphis.Client/Producer/MemphisProducer.cs:
```diff
@@ -77,6 +82,49 @@ public MemphisProducer(MemphisClient memphisClient, string producerName, string
+                    {
+                        sb.AppendLine(error.ToString());
+                    }
+                    throw new MemphisException(sb.ToString());
```
MemphisSchemeValidationException may look better rather generic: MemphisException

elchinmemphis commented:
Overall looks good to me, I left one comment.

elchinmemphis commented:
@bazen-teklehaymanot , I also implemented graphql validation feature, and would like you also review my changes and share your thoughts. [PR](https://github.com/memphisdev/memphis.net/pull/9)

I tried to follow a way such that validators being more testable and extendible,
 
Just created ISchemaValidator and appropriate implementations: GaphqlValidator which being responsible of validation graphql messages.

I believe in that in this way, we can easily write tests for each type of validators and extends when new message type validators needed.

What do you think ?

bazen-teklehaymanot commented:
Sounds good and looking at the changes, it looks more organized now @elchinmemphis. Once, you merge your changes I'll update the JSON validator. 
One question: is there a reason for some of the classes to be public? for example the validators.

elchinmemphis commented:
Thanks!

In the before, Yes, I made SDK specific classes access modifier internal, so that nobody can access and build some logic on that.
But when I created separate project for unit tests, to test classes in SDK, need to make it public otherwise unable to access and test it.

bazen-teklehaymanot commented:
Using the `InternalsVisibleTo` property inside project file can of use, It enables exposing internal classes to specific projects(test projects in this case).

Something like:

```
<ItemGroup>
      <InternalsVisibleTo Include="<NameOfTestProject>" />
</ItemGroup>
```

elchinmemphis commented:
@bazen-teklehaymanot Thanks! your solution regarding **InternalsVisibleTo** really helped to test internal components

idanasulinmemphis commented:
@bazen-teklehaymanot looks great from my side as well, @elchinmemphis I just merged your changes into the master branch, so please take a look on the conflicts, and test it is not breaking your last commit, in case everything is fine you can merge it to master

elchinmemphis commented:
Good job, @bazen-teklehaymanot . 
Looks good to me, only thing that I would like to also have unit tests for validators.

As well as adding features into project, I would highly appreciated to have appropriate unit tests for it.

bazen-teklehaymanot commented:
You are right about the unit tests @elchinmemphis. I've changed few things and added test for JsonValidator. I'll push the latest code after couple of hours

bazen-teklehaymanot commented:
@elchinmemphis I've pushed all changes. I'd be happy to know what you think and if further change is needed. is someone working on #3?

elchinmemphis commented:
Good job! @bazen-teklehaymanot 

Not yet, you can take and work freely :) 
But it would be better to mention on issue that you are working on that [#3 ](https://github.com/memphisdev/memphis.net/issues/3)

elchinmemphis commented:
PR looks good for me.
 @idanasulinmemphis Could you please review and merge?

idanasulinmemphis commented:
@elchinmemphis @bazen-teklehaymanot 1 conflict to resolve before then

bazen-teklehaymanot commented:
@idanasulinmemphis the conflict is resolved
